include/nbl/builtin/hlsl/emulated/matrix_t.hlsl

```diff
@@ -68,27 +69,66 @@ DEFINE_MATRIX_TRAITS_TEMPLATE_SPECIALIZATION(3, 4)
 
 #undef DEFINE_MATRIX_TRAITS_TEMPLATE_SPECIALIZATION
 
+template<typename T, int N, int M, int O>
+struct mul_output<emulated_matrix<T, N, M>, emulated_matrix<T, M, O> >
+{
+    using type = emulated_matrix<T, N, O>;
+};
+
+template<typename T, int N, int M>
+struct mul_output<emulated_matrix<T, N, M>, emulated_vector_t<T, N> >
+{
+    using type = emulated_vector_t<T, N>;
+};
+
 namespace cpp_compat_intrinsics_impl
 {
 template<typename T, int N, int M>
 struct transpose_helper<emulated_matrix<T, N, M> >
 {
     using transposed_t = typename matrix_traits<emulated_matrix<T, N, M> >::transposed_type;
 
-	static transposed_t transpose(NBL_CONST_REF_ARG(emulated_matrix<T, N, M>) m)
+	static transposed_t __call(NBL_CONST_REF_ARG(emulated_matrix<T, N, M>) m)
 	{
         return m.getTransposed();
 	}
 };
 
+template<typename ComponentT, int N, int M, int O>
+struct mul_helper<emulated_matrix<ComponentT, N, M>, emulated_matrix<ComponentT, M, O> >
+{
+    using LhsT = emulated_matrix<ComponentT, N, M>;
+    using RhsT = emulated_matrix<ComponentT, M, O>;
+    using OutputT = typename mul_output<LhsT, RhsT>::type;
+
+    static inline OutputT __call(LhsT lhs, RhsT rhs)
+    {
+        typename matrix_traits<RhsT>::transposed_type rhsTransposed = rhs.getTransposed();
+        const uint32_t outputRowCount = matrix_traits<OutputT>::RowCount;
+        const uint32_t outputColumnCount = matrix_traits<OutputT>::ColumnCount;
+        using OutputVecType = typename matrix_traits<OutputT>::row_type;
+
+        nbl::hlsl::array_set<OutputVecType, typename vector_traits<OutputVecType>::scalar_type> setter;
+
+        OutputT output;
+        for (int r = 0; r < outputRowCount; ++r)
+        {
+            for (int c = 0; c < outputColumnCount; ++c)
+                setter(output.rows[r], c, dot<OutputVecType>(lhs.rows[r], rhsTransposed.rows[c]));
+        }
```

Comment:
this is actually the suboptimal way to perform matrix multiplication (on the CPU, GPU doesn't care).

Best is if you take the rows of LHS and do a weighted sum of them using individual elements of RHS to produce a row of the output, the CPU can still vectorize if it wants to with SSE/AVX
```cpp
for (uint32_t rO=0; rO<OutputRowCount; rO++)
{
   output.rows[rO] = lhs.rows[rO][0]*rhs.rows[0];
   for (uint32_t rI=1; rI<RHSRowCount; rI++) // its also the LHS column count
      output.rows[rO] += lhs.rows[rO][rI]*rhs.rows[rI];
}
``